draft-ietf-quic-transport.md

```diff
@@ -1224,41 +1225,77 @@ be routed back to the endpoint and identified by the endpoint upon receipt.
+NEW_CONNECTION_ID frame ({{frame-new-connection-id}}).
+
+
+### Issuing Connection IDs
+
+An endpoint issues connection IDs to its peer for the peer to use when sending
+packets. This allows the endpoint control over the strategy used to interpret
+connection IDs of packets that it receives.  These connection IDs are
+communicated to the peer using NEW_CONNECTION_ID frames
+({{frame-new-connection-id}}).
+
+The initial connection ID issued to the peer is sent by the endpoint as the
+Source Connection ID during the handshake, ensuring that the peer always begins
+the connection with at least one connection ID to use when sending.  An endpoint
+SHOULD supply its peer with additional connection IDs via NEW_CONNECTION_ID
+frames. While each endpoint can choose how many connection IDs to issue, a
+recommended number of outstanding connection IDs is eight.
+
+When an endpoint issues a connection ID, it MUST accept packets using this
+connection ID for the duration of the connection or until its peer invalidates
+the connection ID via a CONNECTION_ID_FINISHED frame
+({{frame-connection-id-finished}}).
+
+Implementations SHOULD ensure that peers have sufficient connection IDs
+available to reduce the possibility of peers exhausting their supply of
+available connection IDs.  An implementation could do this by always supplying a
+new connection ID for each connection ID retired with a CONNECTION_ID_FINISHED
+frame. When a receiver of a packet notices that its peer is now using a
+previously unused connection ID, it can choose to supply its peer with a new
+connection ID using a NEW_CONNECTION_ID frame to reduce the possibility of its
+peer running out of available connection IDs.
+
+An endpoint that receives a packet with a different remote address than
+previously used SHOULD also switch to sending with a different connection ID.
+This can help to ensure that different connection IDs will be used in both
+directions when an endpoint migrates to a new path or changes connection ID on
+an existing path.
+
+
+### Consuming Connection IDs
+
+If an endpoint's peer has selected a non-zero-length connection ID, the endpoint
+maintains a set of connection IDs received from the peer that it can use when
+sending packets.  All connection IDs issued by the peer are considered valid for
+use by the endpoint when sending packets until the connection ID is retired by
+the endpoint.  Endpoints can choose to stop using a given connection ID to send
+packets at any time and signal this to the issuing endpoint via a
+CONNECTION_ID_FINISHED frame.  This frame indicates the connection ID that is no
+longer in use and serves as a request for the peer to issue additional
+connection IDs via a NEW_CONNECTION_ID frame.
+
+An endpoint that retires a connection ID should retain knowledge of that
+connection ID for a reasonable time after sending the CONNECTION_ID_FINISHED
+frame, or until that frame is acknowledged.  A recommended time is three times
```

Comment:
The acknowledgment point here is a tricky one.  The risk you describe is that a retransmission of NEW_CONNECTION_ID(X) arrives after RETIRE_CONNECTION_ID(X) is sent, which makes it appear as though X is a wholly new connection ID.  If it is then used, it might trigger stateless reset.  Defense against that is to wait for the ACK of the ACK for NEW_CONNECTION_ID, or to start the timer from when NEW_CONNECTION_ID is received, not when RETIRE_CONNECTION_ID is sent.

However, based on comments below, I think that we should recommend retention of the connection ID at the peer that uses it for 3xRTO as well and NOT short that timer when the RETIRE_CONNECTION_ID frame is acknowledged.  The reason for that is not use of the ID (that should have stopped before sending the frame), but acceptance of the stateless reset.  Packets in flight before the connection ID is retired might generate a stateless reset.

The risk I'm more concerned about is the receipt of delayed packets with X after RETIRE_CONNECTION_ID(X) is *received*.  In which case the recipient of RETIRE_CONNECTION_ID needs to remember X and not generate a stateless reset for some period.

That keeps it simple: remember a connection ID (and its associated reset token) from the time of sending or receiving RETIRE_CONNECTION_ID plus 3xRTO.

Reply:
Ah, GitHub was hiding this comment initially.  Remembering retired CIDs for 3xRTO on both sides (accepting SRs, but not generating them) is certainly another solution -- it's more state for the endpoints, but causes Stateless Reset to tear down more quickly.

Reply:
@martinthomson do we still have a reset risk if the client retires the stateless reset value that it received with the retired connection ID?